### PR TITLE
homed: add "recovery key" concept plus track "dirty" state of LUKS volumes

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by poettering** [comment]
BTW, this is on top of #16762, so it might be best to review/merge that one first

File: src/home/meson.build [hunk 1]
```diff
@@ -100,3 +108,11 @@ if conf.get('ENABLE_HOMED') == 1
         install_data('homed.conf',
                      install_dir : pkgsysconfdir)
 endif
+
+tests += [
+        [['src/home/test-modhex.c',
+          'src/home/modhex.c',
+          'src/home/modhex.h'],
+         [libshared],
```
Comment (keszybz) [inline src/home/meson.build]: IIRC, libshared doesn't need to be specified.

File: src/home/homework.c [hunk 2]
```diff
@@ -78,6 +78,26 @@ int user_record_authenticate(
...
+        /* Similar, but test against the recovery keys */
+        r = user_record_test_recovery_key(h, secret);
+        if (r == -ENOKEY)
+                need_recovery_key = true;
+        else if (r == -ENXIO)
+                log_debug_errno(r, "User record has no recovery keys, plaintext passwords not tested against it.");
+        else if (r < 0)
+                return log_error_errno(r, "Failed to validate recovery key of record: %m");
+        else {
+                log_info("Provided password is recovery key that unlocks user record.");
```
Comment (keszybz) [inline src/home/homework.c]: .... is a recovery key that unlocks the user record?

File: src/home/homework.c [hunk 3]
```diff
@@ -78,6 +78,26 @@ int user_record_authenticate(
...
         }
 
+        /* Similar, but test against the recovery keys */
+        r = user_record_test_recovery_key(h, secret);
+        if (r == -ENOKEY)
+                need_recovery_key = true;
+        else if (r == -ENXIO)
+                log_debug_errno(r, "User record has no recovery keys, plaintext passwords not tested against it.");
+        else if (r < 0)
+                return log_error_errno(r, "Failed to validate recovery key of record: %m");
```
Comment (keszybz) [inline src/home/homework.c]: ... to validate the recovery key of the record?

File: src/home/homework.c [hunk 4]
```diff
@@ -914,6 +936,48 @@ static int user_record_compile_effective_passwords(
...
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to test plain text recovery key: %m");
+                        if (r > 0) {
+                                if (ret_effective_passwords) {
+                                        r = strv_extend(&effective, p);
+                                        if (r < 0)
+                                                return log_oom();
+                                }
+
+                                log_debug("Found plain text recovery key.");
```
Comment (keszybz) [inline src/home/homework.c]: "plaintext" spellling is used in other places.

File: man/homectl.xml [hunk 5]
```diff
@@ -376,6 +376,16 @@
         discussion see above.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--recovery-key=</option><replaceable>BOOL</replaceable></term>
+
+        <listitem><para>Accepts a boolean argument. If enabled a recovery key is configured for the
+        account. A recovery key is a computer generated access key that may be used to regain access to an
+        account if the password has been forgotten or the authentication token lost. The key is generated and
+        shown on screen, and should be printed or otherwise transferred to a secure location. They recovery
```
Comment (keszybz) [inline man/homectl.xml]: A recovery key may be entered instead of a regular password to unlock the account.

File: src/shared/user-record-show.c [hunk 6]
```diff
@@ -14,6 +14,8 @@
 const char *user_record_state_color(const char *state) {
         if (STR_IN_SET(state, "unfixated", "absent"))
                 return ansi_grey();
+        else if (streq(state, "degraded"))
```
Comment (keszybz) [inline src/shared/user-record-show.c]: I'd prefer "dirty" here. "degraded" suggests something much more serious. "dirty" is also used everywhere in the code, so I think it'd be less confusing to use the same name here. Also, it shouldn't be red — it should be a recoverable condition, so yellow seems more appropriate.

**Review by keszybz — COMMENTED** [review]
I didn't do a full review — just the changed parts. Looks nice. See some suggestions below. We can merge this as is, and then maybe tweak some things separately if you prefer.

**Comment by poettering** [comment]
force pushed a new version. only changes are the ones @keszybz pointed out. Taking liberty to upgrade green label.

**Pull request merged**